src/docfx/Models/WatchCommand.cs

```diff
@@ -0,0 +1,179 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Docfx.Common;
+using Spectre.Console.Cli;
+
+namespace Docfx;
+
+internal class WatchCommand : Command<WatchCommandOptions>
+{
+    public override int Execute(CommandContext context, WatchCommandOptions settings)
+    {
+        return CommandHelper.Run(settings, () =>
+        {
+            var (config, baseDirectory) = Docset.GetConfig(settings.ConfigFile);
+            BuildCommand.MergeOptionsToConfig(settings, config.build, baseDirectory);
+            var conf = new BuildOptions();
+            var serveDirectory = RunBuild.Exec(config.build, conf, baseDirectory, settings.OutputFolder);
+
+            void onChange()
+            {
+                RunBuild.Exec(config.build, conf, baseDirectory, settings.OutputFolder);
+            }
+
+            if (settings is { Serve: true, Watch: true })
+            {
+                using var watcher = Watch(baseDirectory, config.build, onChange);
+                Serve(serveDirectory, settings.Host, settings.Port, settings.OpenBrowser, settings.OpenFile);
+            }
+            else if (settings.Watch)
+            {
+                using var watcher = Watch(baseDirectory, config.build, onChange);
+
+                // just block but here we can't use the host mecanism
+                // since we didn't start the server so use console one
+                using var canceller = new CancellationTokenSource();
+                Console.CancelKeyPress += (sender, args) => canceller.Cancel();
+                Task.Delay(Timeout.Infinite, canceller.Token).Wait();
+            }
+            else if (settings.Serve)
+            {
+                RunServe.Exec(serveDirectory, settings.Host, settings.Port, settings.OpenBrowser, settings.OpenFile);
+            }
+            else
+            {
+                onChange();
+            }
+        });
+    }
+
+    internal void Serve(string serveDirectory, string host, int? port, bool openBrowser, string openFile) {
+        if (CommandHelper.IsTcpPortAlreadyUsed(host, port))
+        {
+            Logger.LogError($"Serve option specified. But TCP port {port ?? 8080} is already being in use.");
+            return;
+        }
+        RunServe.Exec(serveDirectory, host, port, openBrowser, openFile);
+    }
+
+    // For now it is a simplistic implementation, in particular on the glob to filter mappping
+    // but it should be sufficient for most cases.
+    internal static IDisposable Watch(string baseDir, BuildJsonConfig config, Action onChange)
+    {
+        FileSystemWatcher watcher = new(baseDir)
+        {
+            IncludeSubdirectories = true,
+            NotifyFilter = NotifyFilters.Attributes | NotifyFilters.Size | NotifyFilters.FileName |
+                           NotifyFilters.DirectoryName | NotifyFilters.LastWrite
+        };
+
+        if (WatchAll(config))
```

Comment:
We can start with a simple deterministic watch implementation that watches all files changes in the `docfx.json` directory, except for the output directory (`_site`)